pkg/agent/multicluster/pod_route_controller.go

```diff
@@ -0,0 +1,306 @@
+	ofClient        openflow.Client
+	interfaceStore  interfacestore.InterfaceStore
+	nodeConfig      *config.NodeConfig
+	queue           workqueue.RateLimitingInterface
+	podInformer     cache.SharedIndexInformer
+	podLister       corelisters.PodLister
+	gwInformer      cache.SharedIndexInformer
+	gwLister        mclisters.GatewayLister
+	activeGWName    string
+	namespace       string
+	mutex           sync.Mutex
+	installedPods   map[string]string
+}
+
+type endpointChangeInfo struct {
+	podIP     string
+	podNodeIP string
+}
+
+func NewMCPodRouteController(
+	mcClient mcclientset.Interface,
+	podInformer cache.SharedIndexInformer,
+	gwInformer cache.SharedIndexInformer,
+	client openflow.Client,
+	ovsBridgeClient ovsconfig.OVSBridgeClient,
+	interfaceStore interfacestore.InterfaceStore,
+	nodeConfig *config.NodeConfig,
+	namespace string,
+) *MCPodRouteController {
+	controller := &MCPodRouteController{
+		mcClient:        mcClient,
+		ovsBridgeClient: ovsBridgeClient,
+		ofClient:        client,
+		interfaceStore:  interfaceStore,
+		nodeConfig:      nodeConfig,
+		queue:           workqueue.NewNamedRateLimitingQueue(workqueue.NewItemExponentialFailureRateLimiter(minRetryDelay, maxRetryDelay), "MCPodRouteController"),
+		podInformer:     podInformer,
+		podLister:       corelisters.NewPodLister(podInformer.GetIndexer()),
+		gwInformer:      gwInformer,
+		gwLister:        mclisters.NewGatewayLister(gwInformer.GetIndexer()),
+		namespace:       namespace,
+		installedPods:   make(map[string]string),
+	}
+
+	controller.gwInformer.AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: func(cur interface{}) {
+				controller.enqueueGateway(cur, false)
+			},
+			// Gateway UPDATE event doesn't impact Pod flows, so ignore it.
+			DeleteFunc: func(old interface{}) {
+				controller.enqueueGateway(old, true)
+			},
+		},
+		resyncPeriod,
+	)
+
+	controller.podInformer.AddEventHandlerWithResyncPeriod(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: func(cur interface{}) {
+				controller.enqueuePod(cur, false)
+			},
+			UpdateFunc: func(old, cur interface{}) {
+				controller.enqueuePod(cur, false)
+			},
+			DeleteFunc: func(old interface{}) {
+				controller.enqueuePod(old, true)
+			},
+		},
+		resyncPeriod,
+	)
+	return controller
+}
+
+func (c *MCPodRouteController) enqueueGateway(obj interface{}, isDelete bool) {
+	gw, isGW := obj.(*mcv1alpha1.Gateway)
+	if !isGW {
+		klog.ErrorS(errors.New("received unexpected object"), "enqueueGateway can't process event", "obj", obj)
+		return
+	}
+
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+	// Gateway webhook will guarantee that at most one Gateway in a cluster.
+	if isDelete {
+		c.activeGWName = ""
+		if c.nodeConfig.Name != gw.Name {
+			klog.InfoS("This Node is not the deleted Gateway, skip it", "node", c.nodeConfig.Name, "gateway", gw.Name)
+			return
+		}
+
+		for podIP := range c.installedPods {
+			c.queue.Add(&endpointChangeInfo{
```

Comment:
For Gateway delete/add, do we have a more efficient way to delete/add all flows in a batch? We can do this optimization in a separate PR too.

Reply:
We should not watch Pod events if the current Node is not a Gateway. Watching Pods has cost.

Reply:
I feel there is no batch way for Gateway event since it's actually for Pod IP based flow. The Gateway event is for fully sync. We need each Pod info to update flow rules.
There is a method `validatePod` to exclude Pod if it's on a Gateway Node. Do you mean to add a labelSelector to filter Pods?

Reply:
I meant instead of generating a task for each Pod and queue them one by one to be handled by the worker, can we just have a routine to add flows for all Pods in a batch.

Could you answer my second question too?
> We should not watch Pod events if the current Node is not a Gateway. Watching Pods has cost.

Could we fine a way to start watching Pod events after the current Node becomes a Gateway, and stop watching when it is not?

Reply:
Sure, I will check if there is a way to do this.

Reply:
@jianjuns  I have refined Pod flows controller and it's in the latest commit, please help to review. Thanks.